sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala

```diff
@@ -842,6 +842,333 @@ class FileBasedDataSourceSuite extends QueryTest
+
+        val exactSchema = new StructType().add("camelCase", StringType)
+        checkAnswer(spark.read.schema(exactSchema).parquet(path), Row("A"))
+
+        // In case insensitive manner, parquet's column cases are ignored
+        val caseInsensitiveSchema = new StructType().add("camelcase", StringType)
+        checkAnswer(spark.read.schema(caseInsensitiveSchema).parquet(path), Row("A"))
+      }}
+    }
+  }
+  test("SPARK-31116: Select simple parquet with case insensitive and schema pruning disabled") {
+    withSQLConf(
+      SQLConf.CASE_SENSITIVE.key -> "false",
+      SQLConf.NESTED_SCHEMA_PRUNING_ENABLED.key -> "false") {
+      withTempPath { dir => {
+        val path = dir.getCanonicalPath
+
+        // Prepare values for testing specific parquet record reader
+        Seq("A").toDF("camelCase").write.parquet(path)
+
+        val exactSchema = new StructType().add("camelCase", StringType)
+        checkAnswer(spark.read.schema(exactSchema).parquet(path), Row("A"))
+
+        // In case insensitive manner, parquet's column cases are ignored
+        val caseInsensitiveSchema = new StructType().add("camelcase", StringType)
+        checkAnswer(spark.read.schema(caseInsensitiveSchema).parquet(path), Row("A"))
+      }}
+    }
+  }
+
+  test("SPARK-31116: Select nested parquet with case insensitive and schema pruning enabled") {
+    withSQLConf(
+      SQLConf.CASE_SENSITIVE.key -> "false",
+      SQLConf.NESTED_SCHEMA_PRUNING_ENABLED.key -> "true") {
+      withTempPath { dir => {
+        val path = dir.getCanonicalPath
+
+        // Prepare values for testing nested parquet data
+        spark
+          .range(1L)
+          .selectExpr("NAMED_STRUCT('lowercase', id, 'camelCase', id + 1) AS StructColumn")
+          .write.parquet(path)
+
+        val exactSchema = new StructType()
+          .add(
+            "StructColumn",
+            new StructType()
+              .add("lowercase", LongType)
+              .add("camelCase", LongType))
+        checkAnswer(spark.read.schema(exactSchema).parquet(path), Row(Row(0, 1)))
+
+        // In case insensitive manner, parquet's column cases are ignored
+        val innerColumnCaseInsensitiveSchema = new StructType()
+          .add(
+            "StructColumn",
+            new StructType()
+              .add("LowerCase", LongType)
+              .add("camelcase", LongType))
+        checkAnswer(
+          spark.read.schema(innerColumnCaseInsensitiveSchema).parquet(path),
+          Row(Row(0, 1)))
+
+        val rootColumnCaseInsensitiveSchema = new StructType()
+          .add(
+            "structColumn",
+            new StructType()
+              .add("lowercase", LongType)
+              .add("camelCase", LongType))
+        checkAnswer(
+          spark.read.schema(rootColumnCaseInsensitiveSchema).parquet(path),
+          Row(Row(0, 1)))
+      }}
+    }
+  }
+
+  test("SPARK-31116: Select nested parquet with case insensitive and schema pruning disabled") {
```

Comment:
Since this is the same with the above, you can avoid repetition by merging into into the above test like the following.
```scala
Seq("true", "false").foreach { nestedSchemaPruningEnabled =>
  withSQLConf(
    SQLConf.CASE_SENSITIVE.key -> "false",
    SQLConf.NESTED_SCHEMA_PRUNING_ENABLED.key -> nestedSchemaPruningEnabled) {
    withTempPath { dir =>
       // the same test body
    }
  }
}
```